### ISSUE TITLE
Broken database file after app crash in iOS

### ISSUE THREAD

**Issue by tendant** [tekartik/sembast.dart#3]
App crash by some random bug, which cause the database file to be broken. Looks like sambast could not handle broken database file. 

This is the exception I am getting when try to start the app again.
```
2017-10-16 21:32:52.794543-0700 Runner[1871:656406] [VERBOSE-2:dart_error.cc(16)] Unhandled exception:
FormatException: Unexpected character (at character 134)
...4918-8024-33f979b57a71","name":"g"}}{"key":"51b1d43e-3dbd-421c-918f-a5d3...

#0      _ChunkedJsonParser.fail (dart:convert-patch/dart:convert/convert_patch.dart:1361)
#1      _ChunkedJsonParser.parse (dart:convert-patch/dart:convert/convert_patch.dart:857)
#2      _parseJson (dart:convert-patch/dart:convert/convert_patch.dart:29)
#3      JsonDecoder.convert (dart:convert/json.dart:504)
#4      JsonCodec.decode (dart:convert/json.dart:135)
#5      Database.open.<anonymous closure>.<anonymous closure>.<anonymous closure> (package:sembast/src/sembast_database.dart:748)
#6      Stream.forEach.<anonymous closure>.<anonymous closure> (dart:async/stream.dart:793)
#7      _runUserCode (dart:async/stream_pipe.dart:11)
#8      Stream.forEach.<anonymous closure> (dart:async/stream.dart:793)
#9      _rootRunUnary (dart:async/zone.dart:1132)
#10     _CustomZone.runUnary (dart:async/zone.dart:1029)
#11     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931)
#12     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#13     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:257)
#14     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:68)
#15     _EventSinkWrapper.add (dart:async/stream_transformers.dart:15)
#16     _StringAdapterSink.add (dart:convert/string_conversion.dart:268)
#17     _LineSplitterSink._addLines (dart:convert/line_splitter.dart:156)
#18     _LineSplitterSink.addSlice (dart:convert/line_splitter.dart:131)
#19     StringConversionSinkMixin.add (dart:convert/string_conversion.dart:189)
#20     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:120)
#21     _rootRunUnary (dart:async/zone.dart:1132)
#22     _CustomZone.runUnary (dart:async/zone.dart:1029)
#23     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931)
#24     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#25     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:257)
#26     _SinkTransformerStreamSubscription._add (dart:async/stream_transformers.dart:68)
#27     _EventSinkWrapper.add (dart:async/stream_transformers.dart:15)
#28     _StringAdapterSink.add (dart:convert/string_conversion.dart:268)
#29     _StringAdapterSink.addSlice (dart:convert/string_conversion.dart:273)
#30     _Utf8ConversionSink.addSlice (dart:convert/string_conversion.dart:348)
#31     _Utf8ConversionSink.add (dart:convert/string_conversion.dart:341)
#32     _ConverterStreamEventSink.add (dart:convert/chunked_conversion.dart:86)
#33     _SinkTransformerStreamSubscription._handleData (dart:async/stream_transformers.dart:120)
#34     _rootRunUnary (dart:async/zone.dart:1132)
#35     _CustomZone.runUnary (dart:async/zone.dart:1029)
#36     _CustomZone.runUnaryGuarded (dart:async/zone.dart:931)
#37     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:330)
#38     _BufferingStreamSubscription._add (dart:async/stream_impl.dart:257)
#39     _StreamController&&_SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:770)
#40     _StreamController._add (dart:async/stream_controller.dart:641)
#41     _StreamController.add (dart:async/stream_controller.dart:587)
#42     _FileStream._readBlock.<anonymous closure> (dart:io/file_impl.dart:103)
#43     _rootRunUnary (dart:async/zone.dart:1132)
#44     _CustomZone.runUnary (dart:async/zone.dart:1029)
#45     _FutureListener.handleValue (dart:async/future_impl.dart:129)
#46     _Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:635)
#47     _Future._propagateToListeners (dart:async/future_impl.dart:664)
#48     _Future._completeWithValue (dart:async/future_impl.dart:477)
#49     _Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:509)
#50     _rootRun (dart:async/zone.dart:1124)
#51     _CustomZone.run (dart:async/zone.dart:1021)
#52     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:947)
#53     _microtaskLoop (dart:async/schedule_microtask.dart:41)
#54     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50)
```

**Comment by alextekartik** [comment]
Thanks for the report. Indeed there is no corruption strategy but failing. I will look into that. Assuming a single process access the database (that it its restriction), as an append-only format, only the last lines could be corrupted. I need to take care of transaction so that the last group (transaction) of records are discarded on corrupted data. In the worst case, I guess one strategy (an app should never fails) could be to delete/recreate the database but that would scare people if the data is not synced/backup elsewhere but that could be an option. I have not tried on iOS much (flutter once) so I will take a look

**Comment by alextekartik** [comment]
In the meantime, if the delete-if-corrupted is a valid solution below is a solution with sembast 1.4 until I find a better one. Your suggestions are welcome
````
Database db;
try {
  db = await factory.openDatabase(dbPath);
} on FormatException catch (_) {
  await _deleteFile(dbPath);
  db = await factory.openDatabase(dbPath);
}
````

**Comment by tendant** [comment]
Thanks for the help. 

I just download the database file from device. Looks like the file for some reason two lines are concatenated without new line. 

```
{"version":1,"sembast":1}
...
{"key":"706d719d-056a-4b5a-b390-7af4676578f1","store":"todo_items","value":{"cid":"706d719d-056a-4b5a-b390-7af4676578f1","name":"Ef"}}
{"key":"d4d38732-a680-4918-8024-33f979b57a71","store":"todo_items","value":{"cid":"d4d38732-a680-4918-8024-33f979b57a71","name":"g"}}{"key":"51b1d43e-3dbd-421c-918f-a5d382915fef","store":"todo_items","value":{"cid":"51b1d43e-3dbd-421c-918f-a5d382915fef","name":"test"}}
{"key":"64e21c9c-2766-4691-9641-ff7c3c157f5c","store":"todo_items","value":{"cid":"64e21c9c-2766-4691-9641-ff7c3c157f5c","name":"a"}}
```

**Comment by tendant** [comment]
Since the database is append only, it should trivial to fix broken lines. 

It will be acceptable to fix database file by removing most recent broken lines.

**Comment by alextekartik** [comment]
If that's ok for you and if you still have it, can you send me the whole file? Indeed the file should always terminate by a new line so that appends work fine. Can you confirm that the file was not created manually (i.e without the new line), and that only one process created it; I'm checking right now some race condition where such a think could happen

**Comment by alextekartik** [comment]
I published a new version (1.5.0) that I'd like you to try. There is new open mode DatabaseMode.NEVER_FAILS that will discard records once a corrupted records is encountered. It might not suit your needs as you might loose data but I have added also synchronized block that should prevent concurrent access to the file (in the same process). I'm also interested for more information if you encounter again this issue
